libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+    if (strcasecmp(optarg_pair->name, name)) continue;
+
+    // did we have a find_after, then ignore anything up until our find_after match
+    if ((find_after) && (optarg_pair->value == find_after)) {
+      find_after = NULL;
+      continue;
+    } else if (!find_after) {
+      value = optarg_pair->value;
+    }
+  }
+  printf("cli_get_optarg_value exit - returning <%s><%p>\n", name, value);
+  return value;
+}
+
+void cli_int_free_buildmode(struct cli_def *cli) {
+
+  if (!cli || !cli->buildmode) return;
+  //  cli_unregister_tree(cli, cli->commands, CLI_BUILDMODE_COMMAND);
+  cli->mode = cli->buildmode->mode;
+  free_z(cli->buildmode->cname);
+  free_z(cli->buildmode->mode_text);
+  cli_int_free_found_optargs(&cli->buildmode->found_optargs);
+  free_z(cli->buildmode);
+}
+
+int cli_int_enter_buildmode(struct cli_def *cli, struct cli_pipeline_stage *stage, char *mode_text) {
+  struct cli_optarg *optarg;
+  struct cli_command *c;
+  struct cli_buildmode *buildmode;
+
+  if ((!cli) || !(buildmode = (struct cli_buildmode *)calloc(1, sizeof(struct cli_buildmode)))) {
+    cli_error(cli, "Unable to build buildmode mode for command %s", stage->command->command);
+    return CLI_BUILDMODE_ERROR;
+  }
+
+  // clean up any shrapnel from earlier - shouldn't be any but....
+  if (cli->buildmode) {
+    cli_int_free_buildmode(cli);
+  }
+
+  // assign it so cli_int_register_buildmode_command() has something to work with
+  cli->buildmode = buildmode;
+  cli->buildmode->mode = cli->mode;
+  cli->buildmode->transient_mode = cli->transient_mode;
+  if (mode_text) cli->buildmode->mode_text = strdup(mode_text);
+
+  // need this to verify we have all *required* arguments
+  cli->buildmode->command = stage->command;
+
+  // build new *limited* list of commands from this commands optargs
+  for (optarg = stage->command->optargs; optarg; optarg = optarg->next) {
+    // don't allow anything that could redefine our mode or buildmode mode, or redefine exit/cancel
+    if (!strcmp(optarg->name, "cancel") || (!strcmp(optarg->name, "exit"))) {
+      cli_error(cli, "Unable to build buildmode mode from optarg named %s", optarg->name);
+      return CLI_BUILDMODE_ERROR;
+    }
+    if (optarg->flags & (CLI_CMD_ALLOW_BUILDMODE | CLI_CMD_TRANSIENT_MODE)) continue;
+    if (optarg->mode != cli->mode && optarg->mode != cli->transient_mode)
+      continue;
+    else if (optarg->flags & (CLI_CMD_OPTIONAL_ARGUMENT | CLI_CMD_ARGUMENT)) {
+      if ((c = cli_int_register_buildmode_command(cli, NULL, optarg->name, cli_int_buildmode_cmd_cback,
+                                                  optarg->privilege, cli->mode, optarg->help))) {
+        cli_register_optarg(c, optarg->name, CLI_CMD_ARGUMENT | (optarg->flags & CLI_CMD_OPTION_MULTIPLE),
+                            optarg->privilege, cli->mode, optarg->help, optarg->get_completions, optarg->validator,
+                            NULL);
+      } else {
+        return CLI_BUILDMODE_ERROR;
```

Comment:
There's a memory leak here (and other places below). `buildmode` has been allocated.

Reply:
Dug a bit - it isn't a true leak (memory isn't lost), but the failure case would be the command processing loop thinking buildmode is active when it really isn't.  At any rate, will fix - with one of the few uses of goto that I do in normal code....